### DISCUSSION TITLE
Matrix Discrete Empirical Interpolation Method (MDEIM)

### DISCUSSION THREAD

**Discussion by lebeand** [pymor/pymor#1230] [Ideas]
I found the greedy-based EIM and the DEIM. However, these methods are for parameter-dependent vectors, if I understood that right. In my actual simulation, I need a matrix-based EIM (to obtain an affine decomposition for an efficient online/offline phase). 
Is there such an algorithm currently in PyMOR?
The MDEIM I saw in a lot of papers. It would be a nice extension to PyMOR.

**Comment by renefritze** [comment]
Welcome to pyMOR @andre906!

Have a look at the [ei module](https://docs.pymor.org/2020.2.0/generated/pymor.algorithms.html#module-pymor.algorithms.ei) description and the links therein. 
We also have a demo script that employs EI for solving the [burgers equation](https://github.com/pymor/pymor/blob/main/src/pymordemos/burgers_ei.py)

Is that what you're looking for?

**Reply by sdrave** [reply]
No, @andre906 is asking about MDEIM, which is currently not implemented in pyMOR.

**Reply by lebeand** [reply]
Thank you for the fast replies! My goal is to decompose a matrix `A(\mu)` in parameter-independent matrices and the corresponding parameter-dependent coefficients, i.e. `A(\mu)=\sum_{q=1}^Q \theta_q(\mu) A_q` (sorry for the Latex). 
I've tried with the DEIM-algorithm in PyMOR and I think that's not what I want, although it seems to work.
And I have a few other questions:
- When I assemble my `nonlinop_gen` I get a FixedParameterOperator, which has no matrix attribute. How could I get a matrix representation?

- Why is `nonlinop.apply(U, mu=mu)`, a matrix-vector product, the operator evaluation? (I thought of `nonlinop.assemble(mu)`, but there is no solution of the fom involved, so..., I think I'm missing something. Eventually, you could recommend a paper to help my understanding?)

- Why `lgmres` is used as a solver? The system in fom_ei is still linear. Because of the FixedParameterOperator, I think.


My code:
```python
from pymor.basic import *

#problem
domain = RectDomain([[0., 0.], [1., 1.]], top='neumann')
rhs = ConstantFunction(1., 2)
nonlinfun = ExpressionFunction('np.exp(-((x[...,0]-x1[0])**2+(x[...,1]-x2[0])**2)/sig[0])',
                                                                        2, (),
                                                                        parameters={'sig': 1, 'x1': 1, 'x2': 1},
                                                                        name='nonlinfun')
diffusion = LincombFunction([ConstantFunction(1,2), nonlinfun],
                            [ProjectionParameterFunctional('sig'), ExpressionParameterFunctional('1/sig[0]', {'sig': 1})], name='nonlinop')

dirichlet_data = ExpressionFunction('(x[...,1] <= 0.00000001) * 1.', 2, ())

problem = StationaryProblem(domain=domain,
                            rhs=rhs,
                            diffusion=diffusion,
                            parameter_ranges={'sig': (0.05, 0.1), 'x1': (0.4, 0.6), 'x2': (0.4, 0.6)},
                            dirichlet_data=dirichlet_data)

#discretize
fom, fom_data = discretize_stationary_cg(problem, grid_type=RectGrid, diameter=1 / 3)


#construct ei operator
nonlin_set = problem.parameter_space.sample_uniformly(10)
nonlinop = fom.operator.operators[2]
U_nonlinop = fom.solution_space.empty()
for mu in nonlin_set:
    U = fom.solve(mu)
    U_nonlinop.append(nonlinop.apply(U, mu=mu))

nonlinop_dofs, nonlinop_basis, nonlinop_data = deim(U_nonlinop, product=fom.h1_0_semi_product, atol=10 ** (-5))
#nonlinop_dofs, nonlinop_basis, nonlinop_data = ei_greedy(U_nonlinop, error_norm=fom.h1_0_semi_norm, copy=False, atol=10**(-5))

nonlinop_gen = EmpiricalInterpolatedOperator(nonlinop, nonlinop_dofs, nonlinop_basis, triangular=True)

ei_operators = {'operator': LincombOperator([fom.operator.operators[0], fom.operator.operators[1], nonlinop_gen],
                            [fom.operator.coefficients[0], fom.operator.coefficients[1], fom.operator.coefficients[2]]) }
fom_ei = fom.with_(name=f'{fom.name}_ei', **ei_operators)


# full order solution
mu = Mu({'sig':0.09, 'x1':0.45, 'x2':0.59})
U_fom = fom.solve(mu)
U_fom_ei = fom_ei.solve(mu)
fom.visualize((U_fom, U_fom_ei, U_fom-U_fom_ei), separate_colorbars=True)
print((U_fom - U_fom_ei).norm(product=fom.h1_0_semi_product))
```

**Reply by lebeand** [reply]
Ok, I see my issue. The (D)EIM interpolates a functional `g(u(\mu);\mu)`, like in `A(\mu)*u(\mu) + g(u(\mu);\mu) = f(\mu)`. In my code, I'm interpolating `A_3(\mu)u`, which doesn't make sense. Therefore I get a nonlinear system and the `lgmres` is used. 
- Distinguishes PyMOR's EIM the difference between `g(u(\mu);\mu)` and `g(\mu)`? Because in the first case you would need a nonlinear solver, in the second not.

I conclude that in order to decompose `A(\mu)` in `\sum_{q=1}^Q \theta_q(\mu)*A_q` I would need the MDEIM.

My approach in implementing MDEIM would be to flatten the matrix snapshots 'A(\mu_i), i=1,...,m' and apply DEIM. Only the interface for that operator would be missing. 

EDIT: lgmres is an iterative linear solver.

**Reply by sdrave** [reply]
> My approach in implementing MDEIM would be to flatten the matrix snapshots 'A(\mu_i), i=1,...,m' and apply DEIM. Only the interface for that operator would be missing.

To 'flatten' the operator, you could just use `operator.matrix.data` in the case of a sparse `NumpyMatrixOperator`. For external PDE solvers, you would use something similar. (For memory distributed solvers, things would get difficult.) The bigger problem is to quickly compute the coefficients in the resulting affine decomposition.

**Comment by sdrave** [comment]
Hi @andre906, you are right, there is currently no direct support for MDEIM in pyMOR. Depending on your needs, it would be possible to just treat the parameter-dependent matrix as a linear `Operator` and use the available EIM/DEIM algorithms for it. However, for offline/online decomposition, one would still need to implement the `restricted` method for the respective matrix `Operators`, which should be not too much work, however. Mathematically, this approach would mean that you need to assemble entire selected rows of the matrix.

I also have some code for MDEIM lying around. I haven't found the time yet to properly add it to pyMOR, but if you have a concrete project you want to use this for, I could try to work on it in the relatively near future.

**Reply by lebeand** [reply]
Hi, yes PyMOR warns me about that. Could I replace the `restricted` method with something like `.with_ ` like the operators in the upper example? Is there maybe an example in the pymordemos somewhere?

**Reply by sdrave** [reply]
`.with_` can only be used to change values of `__init__` arguments. However, you could just [monkey patch](https://en.wikipedia.org/wiki/Monkey_patch) `NumpyMatrixOperator` to add the method.

**Comment by sdrave** [comment]
Hi @andre906, sorry for the delay. The following code might do what you want, although it is really empirical interpolation of the coefficient function instead of an interpolation of the matrix. The latter would require code to quickly assemble selected entries of the matrix, which would require support by the PDE solver backend. (Local assembly is currently not implemented for the builtin discretizations. It should be fairly easy to add, however.) For the code to work, you need some minor improvements to `ei_greedy` you can find in #1240. If you have further questions/comments, just ask ..

```python
from pymor.basic import *
import numpy as np

set_log_levels({'pymor.discretizers': 'WARN'})

#problem
domain = RectDomain([[0., 0.], [1., 1.]], top='neumann')
rhs = ConstantFunction(1., 2)
nonlinfun = ExpressionFunction('np.exp(-((x[...,0]-x1[0])**2+(x[...,1]-x2[0])**2)/sig[0])',
                                                                        2, (),
                                                                        parameters={'sig': 1, 'x1': 1, 'x2': 1},
                                                                        name='nonlinfun')
diffusion = LincombFunction([ConstantFunction(1,2), nonlinfun],
                            [ProjectionParameterFunctional('sig'), ExpressionParameterFunctional('1/sig[0]', {'sig': 1})], name='nonlinop')

dirichlet_data = ExpressionFunction('(x[...,1] <= 0.00000001) * 1.', 2, ())

problem = StationaryProblem(domain=domain,
                            rhs=rhs,
                            diffusion=diffusion,
                            parameter_ranges={'sig': (0.05, 0.1), 'x1': (0.4, 0.6), 'x2': (0.4, 0.6)},
                            dirichlet_data=dirichlet_data)

#discretize
fom, fom_data = discretize_stationary_cg(problem, grid_type=RectGrid, diameter=1 / 3)


# evaluate nonlinear coefficient function at grid centers
nonlin_set = problem.parameter_space.sample_uniformly(10)
centers = fom_data['grid'].centers(0)
training_data = np.empty((len(nonlin_set), len(centers)))
for i, mu in enumerate(nonlin_set):
    training_data[i] = nonlinfun.evaluate(centers, mu=mu)
training_data = NumpyVectorSpace.from_numpy(training_data)

# use nodal_basis=True so that interpolation_matrix becomes the identity
idx, basis, ei_data = ei_greedy(training_data, nodal_basis=True, copy=True, atol=10**(-5))

# replace the entries of basis with the corresponding operators
nonlinop = fom.operator.operators[2]
operators = [nonlinop.assemble(mu) for mu in nonlin_set]
operator_basis = [LincombOperator(operators, coeffs).assemble() for coeffs in ei_data['coefficients']]
functionals = [GenericParameterFunctional(lambda mu, i=i: nonlinfun.evaluate(centers[i], mu=mu), nonlinfun.parameters)
               for i in idx]

nonlinop_gen = LincombOperator(operator_basis, functionals)

ei_operators = {'operator': LincombOperator([fom.operator.operators[0], fom.operator.operators[1], nonlinop_gen],
                            [fom.operator.coefficients[0], fom.operator.coefficients[1], fom.operator.coefficients[2]]) }
fom_ei = fom.with_(name=f'{fom.name}_ei', **ei_operators)


# full order solution
mu = Mu({'sig':0.09, 'x1':0.45, 'x2':0.59})
U_fom = fom.solve(mu)
U_fom_ei = fom_ei.solve(mu)
fom.visualize((U_fom, U_fom_ei, U_fom-U_fom_ei), separate_colorbars=True)
print((U_fom - U_fom_ei).norm(product=fom.h1_0_semi_product))
```

**Reply by lebeand** [reply]
I think that works, thank you!
